### DISCUSSION TITLE
Enums - (very experimental) new feature

### DISCUSSION THREAD

**Discussion by phoddie** [phoddie/compileDataView#4] [General]
I thought it could be interesting to try to support enumerations. Often when working with binary data there are constants. Enums are a natural way to describe those in C and TypeScript has a [nice enum feature](https://www.typescriptlang.org/docs/handbook/enums.html) which evolves C enums to JavaScript. While CompileDataView doesn't implement the TypeScript version exactly, it does borrow some of the syntax and ideas.

The enum feature is live, so you can [try out](https://phoddie.github.io/compileDataView/) the examples below in your browser.

The goal in supporting enums is to make it easier to go from a data sheet for a hardware component, the C source code of sensor driver, or the specification of a network protocol to JavaScript code. It will be interesting to try to apply this to some real world projects. I have a few ideas...

The basics work as you would expect. You can declare integer enums that look like C. 

```js
enum Foo {
   zero,
   one,
   three = 3,
   four
};
```

The elements of the enum are accessible as properties:

```js
let five = Foo.four + Foo.one;
```

Enumerations are frozen with `Object.freeze` so that they are truly constant and so that they can be stored entirely in flash memory when using the XS JavaScript engine.

Like TypeScript, you values may be strings, boolean, and even floating point.

```js
enum Foo {
   yes = true,
   no = false,
   unsure =  "maybe",
   almostPI: 3.14   
};
```

The value assigned to enumeration properties are evaluated as JavaScript expressions. This can be useful.

```js
enum Masks {
   a = 1 << 31,
   b = 0x00FF0000,
   c = 0b001100,
   notC = ~c
};

enum Foo {
   zero,
   one,
   two = one + one
};

enum Bar {
   four = Foo.two + Foo.two,
};
```

It can be useful to define the length of arrays using constants. To allow that, structures now evaluate array lengths as expressions:

```js
enum People {
    bob,
    ray,
    tegan,
    sara,
    max
};

struct Stats {
    uint16_t milesWalked[People.max];
    uint32_t stairsClimbed[People.max];
};
```

> **Warning**: As noted in the documentation, the parser in CompileDataView is simplistic. It works pretty well for common cases but expect surprises around the edges.

**Comment by andycarle** [comment]
Very nifty... 

This post is some idle thoughts and doodling on how this enum mechanism and CompileDataView could help define data structures for sensor data and automate sensor data retrieval. Fair warning: this is not fully thought out. But it could be interesting...

***

As you note in your post, enums are a natural fit for translating from a datasheet to code. I've written variations on the following a hundred times over the past decade (this example is substantially reduced from my driver for the [LSm9DS1 IMU](https://cdn.sparkfun.com/assets/learn_tutorials/3/7/3/LSM9DS1_Datasheet.pdf)):

```js
const Registers = Object.freeze({
	CTRL_REG8:		0x22,
	CTRL_REG9:		0x23,
	CTRL_REG10:		0x24,
	OUT_X_XL:		0x28,
	OUT_Y_XL:		0x2A,
	OUT_Z_XL:		0x2C
});
```
This object describes the LSM9DS1's register map, where the keys are the canonical names for the registers and the values are the [SMB](http://smbus.org/specs/SMBus_3_1_20180319.pdf) commands/addresses used to access those registers with SMB reads and writes. The "CTRL" registers are input registers used for configuration while the "OUT_?_XL" registers are the output of the LSM9DS1's accelerometer, split into its three axes. 

This certainly translates cleanly to the CompileDataView enum format:

```ts
enum Registers {
    CTRL_REG8 = 0x22,
    CTRL_REG9,
    CTRL_REG10,
    OUT_X_XL = 0x28,
    OUT_Y_XL = 0x2A,
    OUT_Z_XL = 0x2C
}
```

(An aside: it would be super handy to have a shorthand in the enum syntax for "increment the next value by 2 instead of 1." That's very common in register maps due to 16-bit registers.)

### Observation 1: 
In the case of a register map like this, there is a one-to-one correspondence from the enum to a data structure that could hold the values stored in those registers. On the LSM9DS1 itself it's just a chunk of memory (or maybe actual registers... but probably a chunk of memory). But from the sensor driver developer's perspective, it is helpful to think of it somewhat more abstractly. 

As a CompileDataView structure, the data held in those registers (i.e., the sensor's state) would be something like:

```c
struct LSM9DS1State {
    uint8_t ctrl_reg8_value;
    uint8_t ctrl_reg9_value;
    uint8_t ctrl_reg10_value;
    uint16_t out_x_xl_value;
    uint16_t out_y_xl_value;
    uint16_t out_z_xl_value;
}
```

### Observation 2:
Given the enum and a tiny bit more information, you could generate that CompileDataView input struct automatically. All you would need to know is the width of the non-sequential (and final) registers. 

To completely invent some CompileDataView syntax, it could look something like:

```ts
#pragma generateStructsFromEnums(true)

enum Registers {
    CTRL_REG8 = 0x22,
    CTRL_REG9,
    CTRL_REG10,
    OUT_X_XL(2) = 0x28,
    OUT_Y_XL(2) = 0x2A,
    OUT_Z_XL(2) = 0x2C
}

#pragma generateStructsFromEnums(false)
```

the output of which would be:

```js
export const Registers = Object.freeze({
   CTRL_REG8: 34,
   CTRL_REG9: 35,
   CTRL_REG10: 36,
   OUT_X_XL: 40,
   OUT_Y_XL: 42,
   OUT_Z_XL: 44,
});

export class LSM9DS1State extends DataView {
   constructor(data, offset) {
      if (data)
         super(data, offset ?? 0, 10);
       else
         super(new ArrayBuffer(10));
   }
   get ctrl_reg8_value() {
      return this.getUint8(0);
   }
   set ctrl_reg8_value(value) {
      this.setUint8(0, value);
   }
   get ctrl_reg9_value() {
      return this.getUint8(1);
   }
   set ctrl_reg9_value(value) {
      this.setUint8(1, value);
   }
   get ctrl_reg10_value() {
      return this.getUint8(2);
   }
   set ctrl_reg10_value(value) {
      this.setUint8(2, value);
   }
   get out_x_xl_value() {
      return this.getUint16(4, true);
   }
   set out_x_xl_value(value) {
      this.setUint16(4, value, true);
   }
   get out_y_xl_value() {
      return this.getUint16(6, true);
   }
   set out_y_xl_value(value) {
      this.setUint16(6, value, true);
   }
   get out_z_xl_value() {
      return this.getUint16(8, true);
   }
   set out_z_xl_value(value) {
      this.setUint16(8, value, true);
   }
}
```

### Observation 3:
Given an enum that defines an SMB register map (and the bonus information about register widths...) and the I2C bus details of the part you're talking to, it is well-defined how to mirror state between the data structure and the sensor itself. That is, the getters and setters of the `class X extends DataView` generated by CompileDataView could do SMB writes (setters) and SMB reads (getters) while also shadowing those values into the ArrayBuffer. 

That would give you a perpetually up-to-date binary data structure that shadows the most currently observed state of the sensor. You could then bundle that up for concise network/BLE transmission, send it out over serial/1wire/modbus/etc., store it in flash, or whatever you need to do with it.

Put it all together, and you might have CompileDataView input like:

```ts
#pragma generateStructsFromEnums(true)
#pragma generateSMBCommands(21, 22, 0x6B, 600000) //i2c sda, scl, peripheral address, & bus speed

enum Registers {
    CTRL_REG8 = 0x22,
    CTRL_REG9,
    CTRL_REG10,
    OUT_X_XL(2) = 0x28,
    OUT_Y_XL(2) = 0x2A,
    OUT_Z_XL(2) = 0x2C
}

#pragma generateSMBCommands(false)
#pragma generateStructsFromEnums(false)
```

which would generate output kinda (lots of hand-waving here) like:

```js
export const Registers = Object.freeze({
   CTRL_REG8: 34,
   CTRL_REG9: 35,
   CTRL_REG10: 36,
   OUT_X_XL: 40,
   OUT_Y_XL: 42,
   OUT_Z_XL: 44,
});

export class LSM9DS1State extends DataView {
   #smb;

   constructor(data, offset) {
      if (data)
         super(data, offset ?? 0, 10);
       else
         super(new ArrayBuffer(10));
      this.#smb = new SMBus(scl: 21, sda: 22, address: 0x6B, hz: 600000);
   }
   get ctrl_reg8_value() {
      const value = this.#smb.readByte(Registers.CTRL_REG8);
      this.setUint8(0, value);
      return value;
   }
   set ctrl_reg8_value(value) {
      this.#smb.writeByte(Registers.CTRL_REG8, value);
      this.setUint8(0, value);
   }
   ...
   ...
   ...
   get out_z_xl_value() {
      const value = this.#smb.readWord(Registers.OUT_Z_XL)
      this.setUint16(8, value, true);
      return value;
   }
   set out_z_xl_value(value) {
      this.#smb.writeWord(Registers.OUT_Z_XL, value);
      this.setUint16(8, value, true);
   }
}
```

***

Is this a pretty significant deviation from the core goals of your project? Probably. :) But it could be awfully handy...

**Reply by phoddie** [reply]
Believe it not, I roughly considered this sort of situation. Thank you for giving me an excuse to try it. ;)

Rather than converting enums to a structure, you can just define the structure. The enums have some problems -- they don't tell you the actual data size explicitly, which makes gaps a problem. They also don't tell you the type -- so you don't know if an 8 or 16 bit value is signed or unsigned. Using a structure has some other benefits, like being able to use bitfields -- which are pretty common in SMBus registers. So, the structure gives you all the offsets and bit masks. From there you need to read and write from SMBus rather than a DataView. You can do that with `#pragma extends(AnotherView)` (details in the docs). 

Let's work a simple example. Here's a partial definition for our favorite LIS3DH accelerometer:

```c
enum Range {
	G16 = 0b11,
	G8 = 0b10,
	G4 = 0b01,
	G2 = 0b00
};

enum DataRate {
	HZ400 = 0b0111,	//  400Hz
	HZ200 = 0b0110,	//  200Hz
	HZ100 = 0b0101,	//  100Hz
	HZ50 = 0b0100,		//   50Hz
	HZ25 = 0b0011,		//   25Hz
	HZ10 = 0b0010,		// 10 Hz
	HZ1 = 0b0001		// 1 Hz
};

#pragma extends(SMBusView)
struct LIS3DHView {
	uint8_t  	autoincrement[128];

	uint8_t		pad0[7];
	uint8_t 	STATUS1;
	uint16_t 	OUTADC1;
	uint16_t 	OUTADC2;
	uint16_t 	OUTADC3;
	uint8_t 	INTCOUNT;
	uint8_t 	WHOAMI;
	uint8_t		pad1[0x0F];		// offset 0x10
	uint8_t		TEMPCFG;

//	uint8_t		CTRL1;
	Uint		Xen:1;
	Uint		Yen:1;
	Uint		Zen:1;
	Uint		LPen:1;
	Uint		dataRate:4;

	uint8_t		CTRL2;
	uint8_t		CTRL3;

//	uint8_t		CTRL4;
	Uint		sim:1;
	Uint		st:2;
	Uint		hr:1;
	Uint		fs:2;
	Uint		ble:1;
	Uint		bdu:1;

	uint8_t		CTRL5;
	uint8_t		CTRL6;
	uint8_t		REFERENCE;
	uint8_t		STATUS2;
	int16_t		OUT_X;
	int16_t		OUT_Y;
	int16_t		OUT_Z;
	uint8_t		FIFOCTRL;
	uint8_t		FIFOSRC;
	uint8_t		INT1CFG;
	uint8_t		INT1SRC;
	uint8_t		INT1THS;
	uint8_t		INT1DUR;
	uint8_t		pad2[4];

	uint8_t		CLICKCFG;
	uint8_t		CLICKSRC;
	uint8_t		CLICKTHS;
	uint8_t		TIMELIMIT;
	uint8_t		TIMELATENCY;
	uint8_t		TIMEWINDOW;
	uint8_t		ACTTHS;
	uint8_t		ACTDUR;
};
```

A few things to note:

- The LIS3DH requires you to set 0x80 on the register to auto-increment. There's a 128 byte pad at the front of the structure to set that for every access (which seems safe) . A big strange, but effective.
- The CTRL1 and CTRL4 registers are replaced with bitfields. I think it should be possible to use unions (maybe with a bit of work) to support both full register and bitfield access. But, this works for now.
- It uses `SMBusView` for the `extends` pragma. This means that the generated code extends `SMBusView` instead of `DataView`.

Here's a quick implementation of `SMBusView`:

```js
import SMBus from "pins/smbus";

class SMBusView extends SMBus {
	getUint8(offset) {
		return super.readByte(offset);
	}
	setUint8(offset, value) {
		return super.writeByte(offset, value);
	}
	getUint16(offset) {
		return super.readWord(offset);
	}
	setUint16(offset, value) {
		return super.writeWord(offset, value);
	}
	getInt16(offset) {
		let result = super.readWord(offset);
		if (0x8000 & result)
			return -(65536 - result);
		return result;
	}
	setInt16(offset, value) {
		return super.writeWord(offset, value);
	}
}
```

If you run CompileDataView on the description above and add `SMBusView` to the sources, you can access the sensor very concisely. This is mostly the LIS3DH [example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/drivers/lis3dh/main.js) and [driver](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/drivers/lis3dh/lis3dh.js) (minus interrupt support) from the Moddable SDK:

```js
let sensor = new LIS3DHView({address: 0x19, sda: 21, scl: 22});
if (0x33 !== sensor.WHOAMI)
	throw new Error("unexpected device ID");

sensor.dataRate = DataRate.HZ10;
sensor.fs = 1;
sensor.hr = Range.G4;
sensor.bdu = 1;

const multiplier = 1 / 8190;

Timer.repeat(() => {
	trace(`x: ${sensor.OUT_X * multiplier}, y: ${sensor.OUT_Y * multiplier}, z: ${sensor.OUT_Z * multiplier}\n`);
}, 100);
```

Notice that the bitfields allow access to fields within a register without masking and shifting.

You have access to every obscure feature of the sensor. For example, to disable the X axis of the accelerometer:

```js
sensor.Xen = 0;
```

FWIW - I'm not sold on the shadow idea. There's a lot of overhead to keep it up-to-date. Despite that, it isn't always up-to-date since values on the sensor change independently of being accessed. I would try using the JSON support to serialize the fields you want. It isn't quite what you describe, but maybe close enough to start.
